pkg/v2/tkr/controller/status/clusterstatus/clusterstatus.go

```diff
@@ -0,0 +1,205 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package clusterstatus provides the reconciler for the Cluster UpdatesAvailable status controller.
+package clusterstatus
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	utilversion "k8s.io/apimachinery/pkg/util/version"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	runv1 "github.com/vmware-tanzu/tanzu-framework/apis/run/v1alpha3"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v1/util/patchset"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v2/tkr/resolver"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v2/tkr/util/resolution"
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v2/tkr/util/topology"
+)
+
+type Reconciler struct {
+	Log         logr.Logger
+	Client      client.Client
+	TKRResolver resolver.CachingResolver
+	Context     context.Context
+}
+
+var hasTKRLabel = func() predicate.Predicate {
+	selector, _ := labels.Parse(runv1.LabelTKR)
+	return predicate.NewPredicateFuncs(func(o client.Object) bool {
+		return selector.Matches(labels.Set(o.GetLabels()))
+	})
+}()
+
+const indexCanUpdateToVersion = ".index.canUpdateToVersion"
+
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := mgr.GetFieldIndexer().IndexField(
+		r.Context, &clusterv1.Cluster{}, indexCanUpdateToVersion, versionsClusterCanUpdateTo,
+	); err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("cluster_updates_available").
+		For(&clusterv1.Cluster{}, builder.WithPredicates(hasTKRLabel, predicate.LabelChangedPredicate{})).
+		Watches(
+			&source.Kind{Type: &runv1.TanzuKubernetesRelease{}},
+			handler.EnqueueRequestsFromMapFunc(r.clustersUpdatingToTKR),
+			builder.WithPredicates(predicate.ResourceVersionChangedPredicate{})).
+		Complete(r)
+}
+
+func versionsClusterCanUpdateTo(o client.Object) []string {
+	cluster := o.(*clusterv1.Cluster)
+	if !conditions.IsTrue(cluster, runv1.ConditionUpdatesAvailable) {
+		return nil
+	}
+	message := conditions.GetMessage(cluster, runv1.ConditionUpdatesAvailable)
+	return strings.Split(strings.TrimSuffix(strings.TrimPrefix(message, "["), "]"), " ")
+}
+
+func (r *Reconciler) clustersUpdatingToTKR(o client.Object) []ctrl.Request {
```

Comment:
nit: It would nice to have comment stating how we are using `indexCanUpdateToVersion` indexing and trigger the reconcilation of those clusters which has the TKR listed  in UpdatesAvailable condition gets modified.

Reply:
Done!